server/src/transport/ws.rs

```diff
@@ -106,37 +107,50 @@ where
 // request in the batch and read the results off of a new channel, `rx_batch`, and then send the
 // complete batch response back to the client over `tx`.
 #[instrument(name = "batch", skip(b), level = "TRACE")]
-pub(crate) async fn process_batch_request<L: Logger>(b: Batch<'_, L>) -> BatchResponse {
+pub(crate) async fn process_batch_request<L: Logger>(b: Batch<'_, L>) -> Option<BatchResponse> {
 	let Batch { data, call } = b;
 
-	if let Ok(batch) = serde_json::from_slice::<Vec<Request>>(&data) {
-		return if !batch.is_empty() {
-			let batch = batch.into_iter().map(|req| Ok((req, call.clone())));
-			let batch_stream = futures_util::stream::iter(batch);
-
-			let max_response_size = call.max_response_body_size;
-
-			let batch_response = batch_stream
-				.try_fold(
-					BatchResponseBuilder::new_with_limit(max_response_size as usize),
-					|batch_response, (req, call)| async move {
-						let response = execute_call(req, call).await;
-						batch_response.append(response.as_inner())
-					},
-				)
-				.await;
-
-			match batch_response {
-				Ok(batch) => batch.finish(),
-				Err(batch_err) => batch_err,
+	if let Ok(batch) = serde_json::from_slice::<Vec<&JsonRawValue>>(&data) {
+		let mut got_notif = false;
+		let mut batch_response = BatchResponseBuilder::new_with_limit(call.max_response_body_size as usize);
+
+		let mut pending_calls: FuturesOrdered<_> = batch
```

Comment:
Is there any way to avoid duplicating this logic I wonder? Like a single function whose job it is to take a batch request, run some provided function on each valid item and aggregate the results in the right way.

Maybe more complicated than it's worth, but I wonder how much we could separate out the logic that doesn't care specifically about ws/http and only cares about the formatting and such of messages I guess.

Reply:
yeah, the only thing that's different is the `call` type which needs refactoring to work.

let's do it in another PR eventually :)